### ISSUE TITLE
db:migrate slow with MySQL 8

### ISSUE THREAD

**Issue by bbuchalter** [rails/rails#38523]
### Steps to reproduce
```
git clone git@github.com:bbuchalter/slow_db_migrate_mysql8.git
cd slow_db_migrate_mysql8
bin/demo
```

### Expected behavior
```
BENCHMARK bin/rails db:migrate on MySQL 5.7.29 with no tables setup, no work to do
real	0m2.974s
user	0m2.275s
sys	0m0.937s

Now let's generate 400 tables each with 4 columns and run the migrations.

BENCHMARK bin/rails db:migrate on MySQL 5.7.29 with 400 tables setup, no work to do
real	0m6.669s
user	0m2.782s
sys	0m1.019s
```

### Actual behavior
```
BENCHMARK bin/rails db:migrate on MySQL 8.0.19 with no tables setup, no work to do
real	0m6.788s
user	0m2.967s
sys	0m1.357s

Now let's generate 400 tables each with 4 columns and run the migrations.

BENCHMARK bin/rails db:migrate on MySQL 8.0.19 with 400 tables setup, no work to do
real	0m45.786s
user	0m3.287s
sys	0m1.482s
```

### System configuration
**Rails version**: 6.0.2.1
**Ruby version**: 2.6.5

### Notes
These are the SQL queries that are showing up in the slow query logs:
```
SELECT fk.referenced_table_name AS 'to_table',
       fk.referenced_column_name AS 'primary_key',
       fk.column_name AS 'column',
       fk.constraint_name AS 'name',
       rc.update_rule AS 'on_update',
       rc.delete_rule AS 'on_delete'
FROM information_schema.referential_constraints rc
JOIN information_schema.key_column_usage fk
USING (constraint_schema, constraint_name)
WHERE fk.referenced_column_name IS NOT NULL
  AND fk.table_schema = database()
  AND fk.table_name = 'books_20200219200368'
  AND rc.constraint_schema = database()
  AND rc.table_name = 'books_20200219200368';
```

This seem related to this [change announced in MySQL 8.0](https://dev.mysql.com/doc/relnotes/mysql/8.0/en/news-8-0-0.html):
> Data dictionary tables are invisible, but in most cases there are corresponding INFORMATION_SCHEMA tables that can be queried instead. This enables the underlying data dictionary tables to be changed as server development proceeds, while maintaining a stable INFORMATION_SCHEMA interface for application use.

> Some INFORMATION_SCHEMA tables have been reimplemented entirely as views on data dictionary tables:
```
CHARACTER_SETS
COLLATIONS
COLLATION_CHARACTER_SET_APPLICABILITY
COLUMNS
KEY_COLUMN_USAGE
SCHEMATA
STATISTICS
TABLES
TABLE_CONSTRAINTS
VIEWS
```

I'm not sure what's the best way to address this issue.

cc: @jeremycole @eileencodes

**Comment by eileencodes** [comment]
There's not enough information here for us to look into this. We need a test app or script that shows how it's slow. At the moment this could be Rails or it could be MySQL, but we don't have enough information to determine which. It'd be great to get a profile showing that it is Rails. 

In the future please don't cc core members directly unless you think a change one of us made caused the issue. We'll get around to issues when we can.

**Comment by simi** [comment]
@eileencodes there's link to repo and steps to reproduce this issue in "Steps to reproduce" part.

**Comment by bbuchalter** [comment]
> In the future please don't cc core members directly unless you think a change one of us made caused the issue. We'll get around to issues when we can.

My apologies about this. I was trying to route based on my understanding as this as area of your expertise.

**Comment by eileencodes** [comment]
If someone wants to run a profile that'd be swell, I can't install mysql8 on my machine at the moment.

**Comment by bbuchalter** [comment]
> If someone wants to run a profile that'd be swell, I can't install mysql8 on my machine at the moment.

The demo uses docker to keep the steps to reproduce fully isolated.

**Comment by simi** [comment]
I can do a test run and publish my results.

**Comment by bbuchalter** [comment]
> I can do a test run and publish my results.

FWIW, the actual/expected behavior is the output from the steps to reproduce. Would love to get a second look and confirm we can reproduce. 👍

**Comment by kamipo** [comment]
This is already issued to https://bugs.mysql.com/bug.php?id=98238.

Related #37465.

I'm not sure we can address the issue by our side yet, that should be fixed by MySQL side though.

**Comment by bbuchalter** [comment]
Thanks for sharing this research, I had not been able to find it.

> I'm not sure we can address the issue by our side yet

When I enabled slow query logging in my demo, I was surprised by the number of queries we make to `information_schema`. I'm curious if there is an opportunity to cache. 🤔

**Comment by kamipo** [comment]
This will be fixed in MySQL 8.0.21.
